### ISSUE TITLE
[Feature request] Integrate DeepGEMM

### ISSUE THREAD

**Issue by lessw2020** [pytorch/torchtitan#889]
This might accelerate our MoE computation:
https://github.com/deepseek-ai/DeepGEMM

**Comment by lessw2020** [comment]
Well the gemms are very performance but these are inference only. They didn't release the backward portion ala wgrad.
From their issues disc5seems they are considering releasing but in the interim we can't use them for training yet.

**Comment by lessw2020** [comment]
working on triton implementation to support both inference and training. Bf16 version forward in testing now.

**Comment by rbao2018** [comment]
mark👀

**Comment by lessw2020** [comment]
We have landed a forward MG*NG group Gemm for  deepseek inference this week (bf16)...you can run it usingthe generate.py.
 I also have a triton equivalent of deepseek contiguous groupGemm in another pr. This is bf16 but will add fp8 so we can benchmark vs deepseek deepGemm.
For reference, DeepGemm has two versions, contiguous and masked where masked is for decoding.
Ultimately we will compare all these versions and go with the most performant and flexible options. 
There are also multiple additional kernels in progress to accelerate the entire MoE layers. (sorting and permutation for tokens).
More updates over the next two weeks!

**Comment by vwxyzjn** [comment]
Amazing work @lessw2020!

> From their issues discussion it seems they are considering releasing but in the interim we can't use them for training yet.

It seems https://github.com/deepseek-ai/DeepGEMM/pull/95 just merged the training kernels

**Comment by lessw2020** [comment]
Thanks @vwxyzjn for the update!
We have a cleaner version of deepseek now, so we can potentially integrate there, or just jump to mxfp8 directly.

**Comment by heuristicoder** [comment]
Hi,

Is there any update on this?

**Comment by lessw2020** [comment]
hi @ajWithNucleus, 
I'm no longer working on Titan but maybe @tianyu-l or @danielvegamyhre can provide an update if any plans to integrate these training kernels.

**Comment by codestar12** [comment]
I'd also be interested in an update. I'd be happy to take a crack at integrating it and testing it myself if someone would be kind enough to point me in the right direction.

**Comment by codestar12** [comment]
@tianyu-l @danielvegamyhre
